### DISCUSSION TITLE
Discrepancy Between MiXCR CDR3 Sequences and scRNA-seq Data

### DISCUSSION THREAD

**Discussion by RomiGoldner** [milaboratory/mixcr#1485] [Q&A] [answered]
I'm encountering an issue while correlating MiXCR and 10x Genomics scRNA-seq data. After running `mixcr analyze 10x-sc-5gex`, I extracted CDR3 sequences from the `nSeqCDR3` column. However, I can't find these sequences in the corresponding RNA-seq data from the same cell barcodes.

Could MiXCR's processing lead to CDR3 sequences that don't match raw RNA-seq reads?
Any advice on how to approach this discrepancy?

Appreciate any insights or suggestions!

**Comment by mizraelson** [comment]
Hi,
Based on your message, I understand you performed only 5' gene expression analysis and not a V(D)J-enriched library, correct? Where do you look for the CDR3 sequences, is it the raw FASTQ files? How do you identify the CELL barcode? Since MiXCR corrects erroneous barcodes, there's a possibility that a read containing the CDR3 sequence is tagged with a barcode differing from the raw data.

MiXCR processing can't create a CDR3 sequence that was not present in the reads. you can trace the clones back to the original reads if necessary. To do this, follow the instructions below:


First, run the analysis with these additional parameters:

```
mixcr analyze 10x-sc-5gex \
-Malign.parameters.saveOriginalReads=true \
--prepend-export-alignments-field -descrsR1
--prepend-export-alignments-field -descrsR2
input_R1.fastq.gz \
input_R2.fastq.gz \
sampleName
```

Then, export alignments from the .vdjca file generated by the above command:

```
mixcr exportAlignments \
sampleName.vdjca \
sampleName.alignments.tsv
```

In the `sampleName.alignments.tsv` file, you'll find alignments for each successfully aligned read, including headers for these reads. Note that due to fragmentation, not all reads will encompass the CDR3 region. These reads can then be located in the raw FASTQ files by their headers. Not all of these alignments will end up being assembled to clones, but you can manually check that each CDR3 sequence comes from a certain read. 

Let me know if it helps.

**Reply by RomiGoldner** [reply]
Thank you for your guidance.

Yes, I used the raw R1 and R2 FASTQ files for my analysis. This data was taken from the 10x Genomics website and was not enriched for V(D)J sequences. R1 contains the cell barcodes and UMIs, while R2 carries the RNA sequences, following the standard 10x Genomics pipeline.

Following your advice, I executed the suggested MiXCR commands to generate the alignment file. However, I still encounter the same unexpected issue: when I search for sequences from the `targetSequences` and `nSeqCDR3` columns in the original R2 FASTQ file, I cannot find them. Given that MiXCR constructs CDR3 sequences from these input files, this does not make sense.

Can the sequences in the `targetSequences` column be directly traceable in the R2 file?

**Reply by mizraelson** [reply]
Can you please share a record from the alignements file with the `targetSequences `, `nSeqCDR3 `,`descrsR2` columns and a read from the R2 file with the header that matches the `descrsR2` value?

**Reply by RomiGoldner** [reply]
I'm sorry I didn't mention this in the previous comment but I don't have the descrsR2 column, as I got the following error:
"Can't apply `--prepend-export-alignments-field -descrsR1` because `exportAlignments` step is not presented in the preset."
I ran the command without it.

**Reply by mizraelson** [reply]
What version of MiXCR are you running?

**Reply by RomiGoldner** [reply]
MiXCR v4.6.0

**Reply by mizraelson** [reply]
So I suppose have you already run:

```
mixcr analyze 10x-sc-5gex \
-Malign.parameters.saveOriginalReads=true \
input_R1.fastq.gz \
input_R2.fastq.gz \
sampleName
```

Can you try running after:

```
mixcr exportAlignments \
    -descrsR2 \
    sampleName.vdjca \
    sampleName.alignments.tsv
```

**Reply by RomiGoldner** [reply]
Thank you, that worked.

targetSequences: AAACAGAGTTTGTTTCATATCACAGCCCCCCAGCCTGCAGACTCAGCTACCTACCTCTGTGCACGGAACACCGGTAACCAGTTCTATTTT

nSeqCDR3: 
TGTGCACGGAACACCGGTAACCAGTTCTATTTT

descrsR2:
A00519:644:HCNM3DSXY:1:1576:30942:23265 2:N:0:CACCGCACCA+GACTGTCAAT



FASTQ (R2):
@A00519:644:HCNM3DSXY:1:1576:30942:23265 2:N:0:CACCGCACCA+GACTGTCAAT
AAAATAGAACTGGTTACCGGTGTTCCGTGCACAGAGGTAGGTAGCTGAGTCTGCAGGCTGGGGGGCTGTGATATGAAACAAACTCTGTTT

**Reply by mizraelson** [reply, marked as answer]
So, the R2 reads for 10x are in reverse complement orientation, as shown in the image.

<img width="683" alt="image" src="https://github.com/milaboratory/mixcr/assets/18702359/e2020239-c4df-48e6-835e-d9210faebf7c">

However, the sequences of CDR3 and other regions in MiXCR output are presented in the direct orientation of the coding sequence. 

By reverse complementing R2 from your file, we can observe the sequences as follows:
```
reverseR2: AAACAGAGTTTGTTTCATATCACAGCCCCCCAGCCTGCAGACTCAGCTACCTACCTCTGTGCACGGAACACCGGTAACCAGTTCTATTTT
                                                                    |||||||||||||||||||||||||||||||||
nSeqCDR3:                                                           TGTGCACGGAACACCGGTAACCAGTTCTATTTT
```

**Reply by RomiGoldner** [reply]
Thank you very very much for your help. 
This helped a lot.
